### PR TITLE
BitKeeper History Parsing

### PR DESCRIPTION
**Pull request by jas2701**
I have recently been setting up an OpenGrok server and I wrote this to handle some BitKeeper repositories.  I'm not an expert in BitKeeper (or Java for that matter) but I figured this was probably good enough to be useful to someone.

### CODE REVIEW

**Comment by vladak** [comment]
Looks like `DirectoryListingTest` failed. Care to investigate ?

Also, we will need you to agree and arrange for OCA (Oracle Contributor Agreement).

**Comment by jas2701** [comment]
As far as I can tell the bitkeeper tests have run and failed because bitkeeper was not installed, which I assumed would not be a problem because of the conditional run rule, but it seems all the tests are forced to run.

I'll attempt to add bitkeeper to the travis config, but it's not ideal since it isn't currently available as a deb package as far as I'm aware.

**Comment by tarzanek** [comment]
would it make sense to make the bitkeeper tests optional ?
e.g. if bk binary is found, then continue, if not, then skip ?
I wouldn't want to mandate a binary that is not part of official repos ...

**Comment by tarzanek** [comment]
for OCA: http://www.oracle.com/technetwork/community/oca-486395.html
alt. say you accept OCA here for this patch (and in parallel you can continue with above to get a permanent OCA)

**Comment by vladak** [comment]
I am all for enabling Bitkeeper in Travis (even though it makes it a bit convoluted) as it gets us more test coverage. It's strange that the conditional rules did not work.

**Comment by jas2701** [comment]
I have just sent off a signed copy of the OCA and I confirm that I accept it for this patch.

As for the conditional rules, as far as I could gather when I ran the tests locally the tests were forced to run by FORCE_ALL_PROPERTY being set when [RepositoryInstalled.isSatisfied()](https://github.com/OpenGrok/OpenGrok/blob/master/test/org/opensolaris/opengrok/condition/RepositoryInstalled.java#L54) was called.

**Comment by vladak** [comment]
Looking at the `testdata`, I think the convention is to avoid delivering dot files and rather have them stored without the leading dot (see how `.git` and `.hg` directories are treated).

File: src/org/opensolaris/opengrok/history/BitKeeperAnnotationParser.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/BitKeeperAnnotationParser.java]: copyright year is missing here

File: src/org/opensolaris/opengrok/history/BitKeeperHistoryParser.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/BitKeeperHistoryParser.java]: copyright year is missing here

File: src/org/opensolaris/opengrok/history/BitKeeperHistoryParser.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    /**
+     * Process the output of a `bk log` command.
+     *
+     * Each input line should be in the following format:
+     * either
+     *   D FILE\tREVISION\tDATE\tUSER(\tRENAMED_FROM)?
+     * or
+     *   C COMMIT_MESSAGE_LINE
+     *
+     * BitKeeper always outputs the file name relative to its root directory, which is nice for us since that's what
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/BitKeeperHistoryParser.java]: can you make the line shorter with a line break?

File: src/org/opensolaris/opengrok/history/BitKeeperRepository.java [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+ * See LICENSE.txt included in this distribution for the specific
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/BitKeeperRepository.java]: copyright year is missing here

File: src/org/opensolaris/opengrok/history/BitKeeperRepository.java [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+        } else if (parent.equals("This repository has no pull parent.")) {
+            return null;
+        } else {
+            throw new IOException(executor.getErrorString());
+        }
+    }
+
+    /* History Stuff */
+    /*
+     * BitKeeper has independent revisions for its individual files like CVS, but also provides changesets, which is an
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/BitKeeperRepository.java]: can you break the lines to make them shorter? (there are more of those lines in the file)

File: src/org/opensolaris/opengrok/history/BitKeeperTagEntry.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/BitKeeperTagEntry.java]: copyright year is missing here

File: src/org/opensolaris/opengrok/history/BitKeeperTagParser.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+package org.opensolaris.opengrok.history;
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/history/BitKeeperTagParser.java]: copyright year is missing here

File: test/org/opensolaris/opengrok/history/BitKeeperRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+package org.opensolaris.opengrok.history;
```
Comment (tulinkry) [inline test/org/opensolaris/opengrok/history/BitKeeperRepositoryTest.java]: copyright year is missing here

File: testdata/repositories/bitkeeper/BitKeeper/etc/config [hunk 9]
```diff
@@ -0,0 +1,182 @@
+# This is the BitKeeper configuration for this package.
+#
```
Comment (tulinkry) [inline testdata/repositories/bitkeeper/BitKeeper/etc/config]: As mentioned [by vladak](https://github.com/OpenGrok/OpenGrok/pull/1456#issuecomment-288805553) we do not deliver dot dirs in the repositories.

Check these sections:
 - for ant builds:
   - [repository packing](https://github.com/OpenGrok/OpenGrok/blob/master/build.xml#L770-L798)
   - [repository cleanup](https://github.com/OpenGrok/OpenGrok/blob/master/build.xml#L885-L892)
 - for maven (travis) builds
   - [repository packing](https://github.com/OpenGrok/OpenGrok/blob/master/opengrok-indexer/build.xml#L101-L133)
   - [repository cleanup](https://github.com/OpenGrok/OpenGrok/blob/master/opengrok-indexer/build.xml#L185-L192)

**Comment by vladak** [comment]
Merging this in, will fix the dot files separately.

**Pull request merged**